### PR TITLE
Introducing ScalableSymbolicPulse to correctly compare pulses of amp,angle representation

### PR DESCRIPTION
**Pull request by TsafrirA**
### Summary
Qiskit-Terra PR #9002, converted the library symbolic pulses from a complex `amp` representation to a real (float) `amp`,`angle` representation. To overcome the non-unique nature of the new representation (particularly, when negative amp is allowed), this PR introduces a new subclass `ScalableSymbolicPulse` which assumes the `amp,angle` representation, and compares the pulses accordingly.

PR #9257 presented a more general solution for any non-unique representation, but it was debated whether such a general solution is in fact needed.

### Details and comments
Qiskit-Terra PR #9002, converted the library symbolic pulses from a complex `amp` representation to a real (float) `amp`,`angle` representation. To better support several experiments in Qiskit-Experiments, the `amp` parameter is allowed to take both positive and negative values. This decision creates a non-unique representation for the pulses, where a pi shift of `angle` and a sign flip of `amp` have the same effect (the issue obviously also exists for 2*pi shift of `angle`). For example, the following two pulses

```
from qiskit.pulse.library import Gaussian
import numpy as np

g1 = Gaussian(duration=100, sigma=50, amp=-1, angle=0)
g2 = Gaussian(duration=100, sigma=50, amp=1, angle=np.pi)
```

are not the same when equated (because their parameters are different):
```
g1 == g2 # False
```
but they represent the same waveform:
```
np.all(g1.get_waveform().samples == g2.get_waveform().samples) # True
```

This PR sets to correct this situation, by introducing a new subclass `ScalableSymbolicPulse`. This class assumes that the pulse envelope has the form `amp * exp(1j * angle)*F(t,parameters)` where `F` is some function. When two instances of this class are equated, the complex amplitude `amp * exp(1j * angle)` is compared, but the individual parameters `amp,angle` are ignored.

**The comparison is being done using numpy.isclose(), with default values of rtol=1.e-5, atol=1.e-8**, which I think is sufficient**.

 the comparison of the complex amplitude (and the hashing) is carried out after a rounding to 6 decimal places. Considering the expected scale of the complex amplitude, and HW finite precision, this value should be sufficient to distinguish pulses which are really different from one another.

To accommodate the change, QPY format now includes a `class_name` field, which indicates the correct class when loading the pulse. QPY version was bumped to 6 to reflect this. QPY files of older versions assign only library pulses to the new subclass (even if a custom pulse might fit its purpose).

Returning to the example above, the new PR yields:
```
g1 == g2 # True
```

Along the way, the issue of hashing was raised. Because equal objects should have the same hash, this made hashing very problematic (the only sensible solution was to bin the complex amplitude via rounding). To avoid this issue, and because `SymbolicPulse` hashing was only used for `_is_amplitude_valid`, it was decided to remove the hashing option for `SymbolicPulse` and `ScalableSymbolicPulse`. Note that the caching of `_is_amplitude_valid` is based on a hashing which doesn't take into account the complex amp comparison.

### CODE REVIEW

File: qiskit/pulse/library/symbolic_pulses.py
Comment (nkanazawa1989) [inline qiskit/pulse/library/symbolic_pulses.py:552]: This looks strange. This implements a part of logic of subclass in the superclass. This is not standard python inheritance. Since equality here is very simple, I wouldn't say it's a boilerplate code if you duplicate the rest of logic in the subclass.

File: qiskit/pulse/library/symbolic_pulses.py
Comment (nkanazawa1989) [inline qiskit/pulse/library/symbolic_pulses.py:592]: Since you assume amp and angle exist in the equality evaluation, I think it's more straightforward to add `amp` and `angle` (and do typecheck) to constructor and update parameters dictionary to guarantee these parameters exist.

File: qiskit/pulse/library/symbolic_pulses.py
Comment (nkanazawa1989) [inline qiskit/pulse/library/symbolic_pulses.py:639]: Why hash? These pulses are basically mutable and cannot define hash.

**Review by nkanazawa1989 — COMMENTED** [review]
Overall, adding new subclass seems better than assuming a generic logic. Thanks for proposing new implementation with this PR @TsafrirA .

The implementation looks bit complicated. Current implementation seems almost identical to do
```python
class SymbolicPulse:
    def __init__(self, ..., is_scalable=True)
```
without defining a subclass. To be more explicit, I think the constructor of the subclass can take `amp` and `angle` (as `duration` has a dedicated role), and should full-scratch own equality logic.

For QPY, I think it's better to add class name `*.__class__.__name__` in the `SYMBOLIC_PULSE_PACK_V2` structure, rather than adding `is_scalable` property. In future we may add other subclass with different set of parameter names with assigned role. So class name is much flexible (and we can avoid defining the `SYMBOLIC_PULSE_PACK_V3`). I'm fine with bumping the QPY version in this situation.

File: qiskit/pulse/library/symbolic_pulses.py
Comment (TsafrirA) [reply qiskit/pulse/library/symbolic_pulses.py:592]: I thought about this too, but wasn't sure what's the right way to go.
If you think it's better - I'll move it over to the class.

File: qiskit/pulse/library/symbolic_pulses.py
Comment (TsafrirA) [reply qiskit/pulse/library/symbolic_pulses.py:552]: I think you are right. I didn't want to duplicate the code, but it came out strange. I'll change it.

File: qiskit/pulse/library/symbolic_pulses.py
Comment (TsafrirA) [reply qiskit/pulse/library/symbolic_pulses.py:639]: Well, the `SymbolicPulse` class has a hash function, which is triggered by some tests - which failed (compiler.test_disassembler,compiler.test_assembler). To get those tests to pass, I thought it made sense to have the same hash here.

Comment (wshanks) [reply qiskit/pulse/library/symbolic_pulses.py:689]: The `__hash__` checks that all the parameters have been assigned first. Since we can't re-bind parameters on a pulse, the pulse is immutable at this point. All the test failures came from one use case -- the `lru_cache` on `_is_amplitude_valid()` which calculates the `Waveform` to check if it exceeds 1. We could consider doing this cached evaluation a different way if we wanted to avoid using `__hash__` in `SymbolicPulse`.

There are a two other cases where it could be nice to hash pulses but it is not currently done. In the qobj assembler, symbolic pulses that are not supported by the backend are converted to arrays of samples. These sample arrays are stored in a dictionary keyed on the hash. The code could instead hash the symbolic pulse and skip the waveform conversion. Since we mainly target pulse gates now, I don't think this helps too much in practice since there is separate deduplication of pulse gate calibrations in the qobj assembler. That pulse gate deduplication is actually the second use case. Currently, it works by converting the schedule into qobj instructions, casting the instructions into strings, and hashing them. Using more tailored hashing methods for different instructions could work better than hashing strings, but it probably woudn't help too much in practice (Python can build and hash strings pretty quickly and users don't tend to use a ton of pulse gates currently).

Comment (wshanks) [reply qiskit/pulse/library/symbolic_pulses.py:689]: By the way, if we want to support `__hash__` properly, you should override it here. According to the [Python documentation](https://docs.python.org/3/reference/datamodel.html#object.__hash__),

> The only required property is that objects which compare equal have the same hash value

So the hash needs to account for the angle like `__eq__` does  (which might get a bit awkward with floating point error and shifting angle values by 2 pi).

**Comment by wshanks** [comment]
This implementation looks nice to me. I will let @nkanazawa1989 provide the approval.

I just want to mention a couple things about larger implications.

>  In future we may add other subclass with different set of parameter names with assigned role. So class name is much flexible

What kind of future pulse classes are you thinking of? If they implement features mutually exclusive to `ScalabeSymbolicPulse`, then subclasses make sense. If the features might overlap, then subclasses become more awkward than adding features to the base class since a pulse could only be one subclass at a time.

The other point I want to mention is that we should review when we need to check pulse equality. The main use cases that I can think of are the ones I mentioned in my inline comment about `__hash__` -- when dealing with many pulse gates it can be useful to deduplicate the pulses for efficiency in processing them. Beyond that, I can imagine checking pulses for equality coming up, but I don't know that it is totally necessary (I can't think of a good case other than deduplication; maybe in round trip tests of QPY?). We might just reflect and make sure we think it is worth the extra complexity to support checking pulse equality. One of the motivations for this PR was that with the complex amp->amp+angle change in #9002 it became the case that serializing to QPY and deserializing back to Python could produce a pulse that was not equal to its original source which felt weird. Not supporting `__eq__` would be an alternative solution.

File: qiskit/pulse/library/symbolic_pulses.py
Comment (TsafrirA) [reply qiskit/pulse/library/symbolic_pulses.py:689]: Thanks @wshanks for pointing out the missing implementation of the `__hash__`.
I guess the way to do it properly is to use the value of the complex amp with the same level of accuracy for both `__eq__` and `__hash__`.

Comment (jakelishman) [reply qiskit/pulse/library/symbolic_pulses.py]: If there's fuzziness in a floating-point comparison somewhere in `__eq__`, you can't implement `__hash__`.  Technically, the only valid return value from `__hash__` is static (i.e. independent of the input[^1]).  Actually coding that is worse than disabling it, because it makes every `dict` operation the worst-case O(n) instead of amortised constant.

Consider trying to implement `__hash__` for a float subclass that has `def __eq__(a, b): return abs(a - b) <= 2**-10`.  Now the numbers `1` and `1 + (2**-10)` are equal, so must have the same hash, which is also true for every number between them too.  But also, `1 + (2**-9)` is equal to `1 + (2**-10)`, so it has to have the same hash as `1` as well.  We can keep going until `2` must have the same hash as `1`, and so on and so on.[^2]

[^1]: except for special values like `inf` and `nan`
[^2]: strictly the absolute comparison I wrote would actually reduce to exact equality at float values of about `2**43`, I just wanted the maths to be simpler than using proper ULP comparisons.

Comment (TsafrirA) [reply qiskit/pulse/library/symbolic_pulses.py]: Hi @jakelishman ,
Thank you for your comment. I agree that this implementation seems weird. As soon as I pushed this commit I PM’d some people asking them for an input. Obviously, it would have been smart to include you, and better explain myself here.

However, I think under some assumptions it does work (and by all means, correct me if I’m wrong).

First of all, note that I didn’t use a comparison like the one you have in your `__eq__`. I rounded the values before comparing them. For low enough values, the granularity of the round function creates what is essentially a fixed point value. And here comes the first assumption – I assume that the magnitude of the complex amplitude is O(1). By creating a fixed-point-like representation, we have a concrete (numerical uncertainty free) way to both compare and hash.

The second assumption I make, and I really should have posted it here as a question, is that a fixed-point-like representation with high enough resolution is sufficient for all practical purposes – considering the finite precision of the HW. I am guessing `10^-6` is low enough value, but we’d need to figure it if that’s true.
Going back to the example you provided, while `0.4e-7` and `0.3e-7` will be judged equal and hashed the same, for all practical purposes I am guessing they really do behave the same. Note that there is no cascading effect like the one you described, because of the round function.

 I hope this makes my logic clearer. As I said, this does feel weird, but I think it nonetheless works (under the assumptions I provided). I think we should have a discussion about the three options we have – find an agreed upon hashing solution, let go of hashing, or leave in place the non-unique nature of the amp-angle representation.

Comment (jakelishman) [reply qiskit/pulse/library/symbolic_pulses.py]: I haven't read the code in detail for what's being doing, but what you're saying sounds like you're effectively binning the data for certain operations, and maybe not for others. You're right that will let you make a valid hash, but it also needs a lot of caution when using the objects, especially if mathematical operations like addition of pulses do not use the same binning behaviour - you can easily end up in the situation where `a1 == a2` and `b1 == b2`, but `a1 + b1 != a2 + b2`, for example.

If the only valid way to use the objects is an the discretised version of the number line, and what you say here implies that you strongly expect _not_ to use the dynamic range of floating-point values, might things be simpler/clearer for developers and users if you use an integer representation of these with a single floating-point value to store the "bin width" scaling? That makes it harder to combine pulses of different precisions, but I think in the hardware model you're trying to describe, that's impossible anyway.

(I have no idea if that is a sensible idea, I just mention it because it seems to make explicit the model the code you describe implies, and doesn't allow people to forget it.)

On the subject of hashing, my view would generally err on the side of caution if there are _any_ questions about whether it makes sense, is efficient, or a user might reasonably mutate parts of the state that contribute to equality in any situation.

Comment (TsafrirA) [reply qiskit/pulse/library/symbolic_pulses.py]: Yes, you are correct. I effectively bin for some operations and not others. I can't see any immediate and obvious failure point like the examples you described, because there are very few operations you can actually do with the pulses at the moment. However, I agree that this is a risky maneuver. I don't think we can (or want) to bin the data in the pulse itself - firstly because the representation of the pulse is based on the `amp,angle` represenation, while the binning is carried out for the complex amplitude. Furthermore, I suspect actually binning the data would require some more detailed knowledge about the backend, which will limit the scope of Qiskit. In any case, I don't think this is a good way to go just to solve this non-uniqueness issue. If we don't think this is a good enough solution (and I am starting to feel like this is the case), I think we will be forced to ignore this issue, and leave it as a quirk in the code.

**Review by nkanazawa1989 — COMMENTED** [review]
Thanks @TsafrirA recent updates seem reasonable to me, but I also think revisiting the demand for hash is good idea. As @wshanks pointed out this is currently only used to validate amplitude (to avoid calling computationally heavy `.get_waveform` as much as possible). 

If we don't have further motivation for hashing (I think no), we can change `_is_amplitude_valid` to directly call this line and let it take envelope and parameters which are both hashable.
https://github.com/Qiskit/qiskit-terra/blob/2f5944d60140ceb6e30d5b891e8ffec735247ce9/qiskit/pulse/library/symbolic_pulses.py#L490

**Comment by nkanazawa1989** [comment]
> What kind of future pulse classes are you thinking of? If they implement features mutually exclusive to ScalabeSymbolicPulse, then subclasses make sense. If the features might overlap, then subclasses become more awkward than adding features to the base class since a pulse could only be one subclass at a time.

I don't have any concrete idea in my mind, but I was just on the safe side. The original implementation introduced a boolean value to distinguish two classes. If we introduce another subclass in future PR we would need to bump the QPY version again and permanently maintain the another legacy loader. If we save class name instead of boolean we don't need to introduce such complexity.

File: qiskit/qpy/binary_io/schedules.py
Comment (TsafrirA) [reply qiskit/qpy/binary_io/schedules.py:114]: Corrected. Thank you.

File: qiskit/pulse/library/symbolic_pulses.py
Comment (TsafrirA) [reply qiskit/pulse/library/symbolic_pulses.py]: Thank you.

File: qiskit/pulse/library/symbolic_pulses.py
Comment (TsafrirA) [reply qiskit/pulse/library/symbolic_pulses.py]: Eventually, we removed the hashing, and avoided this issue.

**Review by nkanazawa1989 — COMMENTED** [review]
I like new logic for the amplitude validation. Making pulses immutable looks right approach since they can be parameterized. This is almost good to go.

File: qiskit/pulse/library/symbolic_pulses.py
Comment (TsafrirA) [reply qiskit/pulse/library/symbolic_pulses.py:689]: Thanks.

File: qiskit/pulse/library/symbolic_pulses.py
Comment (TsafrirA) [reply qiskit/pulse/library/symbolic_pulses.py:86]: I agree. That's better.

File: qiskit/qpy/binary_io/schedules.py
Comment (TsafrirA) [reply qiskit/qpy/binary_io/schedules.py:135]: You are right. Thank you.

File: releasenotes/notes/new_pulse_subclass-44da774612699312.yaml
Comment (nkanazawa1989) [inline releasenotes/notes/new_pulse_subclass-44da774612699312.yaml:19]:
```suggestion
other:
  - |
    QPY version was bumped to 6. Following the introduction of :class:`~qiskit.pulse.library.ScalableSymbolicPulse`
    the header of saved :class:`~qiskit.pulse.library.SymbolicPulse` objects was modified to include
    a `class_name_size` attribute, and the QPY version was bumped to reflect this.
```

Please move this to upgrade section. This is an example text for version 5 upgrade.
>  The QPY version format version emitted by :func:`.qpy.dump` has been
increased to version 5. This new format version is incompatible with the
previous versions and will result in an error when trying to load it with
a deserializer that isn't able to handle QPY version 5. This change was
necessary to fix support for representing controlled gates properly and
representing non-default control states.

File: qiskit/pulse/library/symbolic_pulses.py
Comment (wshanks) [inline qiskit/pulse/library/symbolic_pulses.py:103]: Just checking -- is it still best to use `Waveform` for this check now that we are calculating the array directly? We could just check if all samples are below 1 (or 1+1e-7 like `Waveform` does).

File: qiskit/pulse/library/symbolic_pulses.py
Comment (wshanks) [inline qiskit/pulse/library/symbolic_pulses.py:95]: This signature needs to be updated.

File: qiskit/pulse/library/symbolic_pulses.py
Comment (wshanks) [inline qiskit/pulse/library/symbolic_pulses.py:557]: Should we note this removal in the release notes?

**Review by wshanks — COMMENTED** [review]
I like the removal of `__hash__` as the path forward. I noted a few small things.

File: qiskit/pulse/library/symbolic_pulses.py
Comment (TsafrirA) [reply qiskit/pulse/library/symbolic_pulses.py:103]: Technically, nothing has changed compared to what we did before. It could be argued that it's best to leave the test logic in a single place, as it is not trivial. I am fine with both options.

Comment (nkanazawa1989) [reply qiskit/pulse/library/symbolic_pulses.py:103]: I agree with Will. We can directly test amplitude limit, which should be slightly faster.

File: qiskit/pulse/library/symbolic_pulses.py
Comment (nkanazawa1989) [reply qiskit/pulse/library/symbolic_pulses.py:557]: Good point. Should be noted as upgrade.

File: qiskit/pulse/library/symbolic_pulses.py
Comment (TsafrirA) [reply qiskit/pulse/library/symbolic_pulses.py:95]: Thank you.

File: qiskit/pulse/library/symbolic_pulses.py
Comment (TsafrirA) [reply qiskit/pulse/library/symbolic_pulses.py:557]: Thank you.

**Review by nkanazawa1989 — APPROVED** [review]
Thanks @TsafrirA this looks good to me.

**Review by wshanks — APPROVED** [review]

**Comment by wshanks** [comment]
@nkanazawa1989 Does one of us put the automerge tag on this or do we let someone from the core team do that? I have this open in a browser tab and each time I pass by it I see it out of date with main and still re-running the checks, so I am not sure a human will ever get a chance to merge it :slightly_smiling_face:

**Comment by jakelishman** [comment]
For reference: Will and I talked offline, and the answer is that code-owners are completely allowed (encouraged, even!) to apply "automerge" themselves when something's appropriate to merge.  It's much better to use the bot than to do it manually because the bot will only update the PR when it's its turn to merge, avoiding unnecessary CI usage to test updates that'll never be up-to-date.

**Comment by nkanazawa1989** [comment]
Thanks Will for adding the automerge tag! I was waiting for your final review this time (if you wanted).

**Comment by wshanks** [comment]
I might as well mention for @nkanazawa1989 and @TsafrirA that Jake also pointed out that the reno file uses reStructuredText format rather than markdown, so for example in line code formatting should use two `` ` `` around the text instead of one.

**Comment by TsafrirA** [comment]
Thanks for the heads up @wshanks . This one's on me.

**Pull request merged**
